### ISSUE TITLE
xarray.Dataset from existing netCDF4.Dataset

### ISSUE THREAD

**Issue by dopplershift** [pydata/xarray#1459]
It would be really handy to be able to initialize a `xarray.Dataset` instance from an already opened instance of `netCDF4.Dataset`. I have a lot of code where I'm already returning an opened netCDF4 file and this would streamline the process of hooking xarray into that.

It seems like the quick solution here would be to make `NetCDF4DataStore` accept a `netCDF4.Dataset` instance as `filename`, which would bypass the creation of a new instance. Thoughts?

**Comment by fmaussion** [comment]
I needed this too and proposed a change which wasn't accepted a while ago: https://github.com/pydata/xarray/pull/1060 (the PR topic changed in the meantime, so it's hard to follow).

I am still using the same quite ugly solution. I have a [shallow subclass](https://github.com/fmaussion/salem/blob/05577bec6cf3a89b3ef9c30aeb43906a4e9b8f99/salem/sio.py#L884) of ``NetCDF4DataStore``  which I then [feed to](https://github.com/fmaussion/salem/blob/05577bec6cf3a89b3ef9c30aeb43906a4e9b8f99/salem/sio.py#L989) ``open_dataset``...

**Comment by shoyer** [comment]
We could add an alternative constructor to `NetCDF4DataStore`. This would work fine in most cases, with one exception: attempting to pickle the data store would fail. But as long as you mention that downside (relevant for using dask distributed) I'd be fine with this.
